pkg/daemon/update.go

```diff
@@ -312,6 +333,38 @@ type MachineConfigDiff struct {
 	units    bool
 }
 
+// See if this was Rust we wouldn't have nullable pointers by default...
+func canonicalizeStringArray(a []string) []string {
+	if a == nil {
+		return []string{}
+	}
+	return a
+}
+
+// NewMachineConfigDiff compares two MachineConfig objects.
+func NewMachineConfigDiff(oldConfig, newConfig *mcfgv1.MachineConfig) *MachineConfigDiff {
+	oldIgn := oldConfig.Spec.Config
+	newIgn := newConfig.Spec.Config
+	return &MachineConfigDiff{
+		osUpdate: oldConfig.Spec.OSImageURL != newConfig.Spec.OSImageURL,
+		kargs:    !reflect.DeepEqual(canonicalizeStringArray(oldConfig.Spec.KernelArguments), canonicalizeStringArray(newConfig.Spec.KernelArguments)),
```

Comment:
you were hitting the issue of nil vs empty slices bc you are using deepequal to compare. you can do my suggestion above. or just check the len of both slices if !=0 then deepequal else true.

Reply:
yeah! Added checks to consider both empty and nil slices of kargs as same and not perform deepcopy() in this case